_posts/2017-10-04-scalamacros.md

```diff
@@ -0,0 +1,299 @@
+approach while adding support for def macros with access to the semantic API.
+The issues we encountered are documented in more detail in section 3 of the
+technical report ["Two Approaches to Portable Macros"].
+
+Nevertheless, we learned a few important lessons
+
+- Users and macro authors care a lot about IDE support, in particular IntelliJ
+  support.
+- Building a portable macro system with ASTs that don't leak compiler
+  implementation details is possible.
+- Eagerly converting between the compiler and scalameta ASTs introduces
+  a lot of problems that can be avoided by implementing macros in terms of
+  extractor/constructor methods on *abstract syntax*.
+
+With these lessons learned, we decided to retire our efforts to build a macro
+system on top of Scalameta.
+From now on, Scalameta's primary focus is to support building developer tools
+such as Scalafmt, Scalafix and Metadoc.
+
+### v3: scala.macros
+
+In winter and spring 2017, Eugene Burmako at Twitter and Liu Fengyun at EPFL
+worked on a new macro system to address the limitations of scala.reflect-based
+and scala.meta-based macros.
+This third iteration of Scala Macros builds on top of the strengths of the
+scala-reflect API with the following distinction, it's
+
+- smaller, the API exposes as little as possible from the compiler
+  while still being able to support most interesting macro applications.
+- more robust, common pitfalls in scala.reflect-based macros are guarded by
+  the type system with separation of typed and untyped trees.
+
+The design of this approach is explained in more detail in
+section 4 "The Syntax-Based approach" of the technical report
+["Two approaches to portable macros"].
+The latest implementation of this design is hosted at [scalacenter/macros].
+I want to thank Twitter for allowing Eugene Burmako to share this
+implementation, that he created during work-hours, under a BSD 3 license.
+This repository has a working prototype that we can base our future work on.
+
+One paint point in macros v1 that macros v3 does not address is the separate
+compilation restriction.
+Macro definitions must still be compiled in a separate project from where they
+are used.
+Nevertheless, we believe that v3 represents a significant enough improvement to
+forgive this restriction, which is admittedly an inconvenience for macro authors
+but arguably not a blocker for adoption.
+
+## Next steps
+
+As the history above shows, establishing a stable macro system for Scala is a
+large undertaking.
+It has taken many years to reach where we are today, involving a collaboration
+between many different parties.
+We still have a long way to go to reach the level of expressiveness, robustness
+and simplicity that we seek in a stable macro system.
+
+Here below is a rough estimated roadmap for macros v3
+
+- in Scala 2.12, we experiment with compiler plugins
+- in Dotty, Liu Fengyun at EPFL will work on integrations
```

Comment:
What does "integrations" mean? That's a bit too ambiguous...

Reply:
Agree, I didn't like this either, is it better now?